Take into account cfs quota and affinity.

sdab opened the pull request:
In addition to cpusets, cpu affinity can constrain cpus so this adds looking at sched_getaffinity in addition to effective cpus in the cgroup.

For the recommended threads, we additionally take into account cfs quota doing similar quota->cpu math as systemd and lxcfs.

tych0 commented on src/lib.rs:
```diff
@@ -1,3 +1,6 @@
+use nix::sched::sched_getaffinity;
+use nix::sched::CpuSet;
+use nix::unistd::Pid;
```
This looks unused?

tych0 replied:
Oh, I see where you call it.

tych0 commented on src/lib.rs:
```diff
@@ -64,31 +93,61 @@ fn parse_effective_cpus(raw_cpus: String) -> Result<c_int> {
+
+fn get_affinity_cpu_count() -> Result<c_int> {
+    let pid = Pid::from_raw(0);
+    let cpuset = sched_getaffinity(pid)?;
```
Can you describe what sched_getaffinity() does that isn't covered by cpusets? I suppose it takes into account isolcpus?

I'm asking because I think we're more likely to be successful in landing this elsewhere if we have less external dependencies.

sdab replied:
They are just totally different mechanisms. cpusets and cpu affinity masks are distinct, its just that the scheduler takes them both into account.

You can see this by setting an affinity mask and seeing that cpuset.cpus.effective doesnt notice or by setting a cpuset and seeing that sched_getaffinity doesnt notice.

But we should take both into account since they can both constrain the scheduler.

tych0 replied:
I see, makes sense, thanks. I wonder if we can call this syscall directly, vs. having an external dependency. but we already have a library size problem, so it's probably fine for now.

tych0 replied:
Looks to me like cpusets are a strict superset of affinity: https://github.com/torvalds/linux/blob/610347effc2ecb5ededf5037e82240b151f883ab/kernel/sched/core.c#L8348-L8349

Does that mean we can get rid of the cgroup querying code (for cpusets at least) all together?

sdab replied:
if you set a cpuset and then call taskset -cp <pid> (which calls sched_getaffinity), it doesnt have any effect.

tych0 replied:
Is that because taskset looks at cpusets (via sched_getaffinity) before it tries to set? IIUC, you could manually sched_setaffinity() and that would work, assuming that set was a subset of your cpuset. (What I'm really asking is: can we delete some of my code in favor of yours here?)

tych0 commented on src/lib.rs:
```diff
@@ -64,31 +93,61 @@ fn parse_effective_cpus(raw_cpus: String) -> Result<c_int> {
     }
 
-    Err(Errno { errno: ENOENT })
+    Ok((quota / period) as c_int)
```
Can you write a little blurb (maybe in the readme or somewhere) about why this is correct? I'd like to have some pseudocode too, so even if people don't use this code directly, they can use it as a reference.

sdab replied:
done

tych0 commented on src/lib.rs:
```diff
@@ -46,15 +55,35 @@ fn my_cgroup() -> Result<PathBuf> {
                 Ok(end.parse::<c_int>()? - start.parse::<c_int>()? + 1)
-            } else if let [single] = &membs[..] {
-                Ok(single.parse::<c_int>()?)
+            } else if let [_single] = &membs[..] {
```
Why _?

sdab replied:
rust was complaining about an unused variable, this was the suggested fix

tych0 replied:
Oh, right. In that case you can just call it `_`, which I think is the idiom for unused.

sdab replied:
done

tych0 commented:
> there was a small bug with cpuset.cpus.effective parsin

Can you make this one a separate commit, and add a test that exercises the bug?

edit: also, looks like these all need signed-off-by.

sdab commented:
> > there was a small bug with cpuset.cpus.effective parsin
> 
> Can you make this one a separate commit, and add a test that exercises the bug?
> 
> edit: also, looks like these all need signed-off-by.

It is a separate commit, do you mean separate PR? The new test cases do exercise the bug

tych0 commented:
> It is a separate commit, do you mean separate PR?

Ideally the bug fix would live in a commit on its own, separate from the tests for the new feature.

sdab commented:
signed commits

sdab committed:
Take into account cfs quota for recommended_threads call.

The way we convert quota to cpu count is by modelling full time on one cfs_period as one cpu. Thus 100% is 1 cpu, 200% is 2 and so on. We return 1 cpu if the quota is <100% as well. This is how systemd models CPUQuota as well as lxcfs.

We use this cpu count to determine how many logical cpus are available for recommended_threads.

Signed-off-by: Sebastien Dabdoub <sebastien.dabdoub@gmail.com>

sdab committed:
Use sched_getaffinity instead of cpusets.cpus.effective to get cpu count.

sched_getaffinity is more general than cpusets, but does take into account cpusets. This does pull in an addiitonal library, but we can perform the syscall ourselves if thats an issue.

Signed-off-by: Sebastien Dabdoub <sebastien.dabdoub@gmail.com>

sdab committed:
Improve testing by actually creating cgroups.

The current set of bats tests create and mount a tempdir and use that as the rootfs for the example job. This doesnt work with sched_getaffinity, so I made them create a temp cgroup and run the example program in there.

In fact, the previous tests didnt even check the output of the example program and thus were always passing.

Signed-off-by: Sebastien Dabdoub <sebastien.dabdoub@gmail.com>

sdab commented:
Ok I re-worked the git commits and switched out effective cpus for sched_getaffinity. PTAL

tych0 commented:
Looks great, thank you!
